### PR TITLE
catalog: validate/repair default privs with orphan roles

### LINKED ISSUE #130243: catalog: support validating / repair orphan roles in default privileges
Clusters created before https://github.com/cockroachdb/cockroach/pull/67950 could have default privileges applied on roles that no longer exist, which were never cleaned up.

Due to this, we can have clusters that keep getting a non-existent role added in its `privileges`. To help repair this error, we are going to add validation and automatic repair for fixing scenarios where a dropped role exists within a default privilege.

Jira issue: CRDB-41953

Epic CRDB-40419

### PR DESCRIPTION
**Pull request by annrpom**
### sql/catalog: check default privs during desc validation

This patch adds an additional step for our invalid objects
virtual table that checks if default privilege descriptors
are invalid. A default privilege descriptor is considered
invalid if it meets at least one of the following:

1. The role that the default privilege is granted *on* does
not exist
2. The role that the default privilege is has been granted *to*
does not exist

Release note: None

---

### sql/catalog: add default priv repair step

This patch adds a repair step to strip away any
non-existent roles in default privileges.

Fixes: #130243

Release note (bug fix): Add automated clean-up/validation
for dropped roles inside of default privileges.

### CODE REVIEW

**Comment by annrpom** [comment]
Some of the logictest configs (fakedist, 3node-tenant, local-read-committed) are failing with `expected results are invalid: unexpected column count 15 != 5 ([106 test2 · · a default privilege exists for a role "roacher" that doesn't exist])`

I don't see how this is related to my PR so I plan on skipping those configs (putting this here in case anyone objects)

**Review by rafiss — CHANGES_REQUESTED** [review]
nice, this looks good so far!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-:-O8IETTC1_tKzKlpnxmD:b-w77w2l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/crdb_internal.go` line 6225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IDIGVAdDio3U7Od7s:-O8IDIGVAdDio3U7Od7t:b-3sizu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/crdb_internal.go#L6225)):*
> ```Go
> 			}
> 			doError(catalog.ValidateRolesInDescriptor(descriptor, roleExists))
> 			if dbDesc, ok := descriptor.(catalog.DatabaseDescriptor); ok {
> ```

schema descriptors can also have default privileges in them; let's check them here too
___
*[`pkg/sql/catalog/dbdesc/database_desc_builder.go` line 187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IEAZpCQvIsEPiQsCD:-O8IEAZpCQvIsEPiQsCE:b-cpn84w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/catalog/dbdesc/database_desc_builder.go#L187)):*
> ```Go
> ) error {
> 	// Remove any non-existent roles from default privileges.
> 	defaultPrivs := ddb.original.GetDefaultPrivileges()
> ```

we'll also need similar logic for schema descriptors
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9ykp7mDBTekCG6bF:-O8I9ykp7mDBTekCG6bG:b4zcf9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1121)):*
> ```
> 
> statement ok
> SET ROLE admin;
> ```

super nit: the default role in logic tests is `root`, so let's use that here.
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IEI-I-kdAn0cs60Po:-O8IEI-I-kdAn0cs60Pp:b-iyklyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1124)):*
> ```
> 
> statement ok
> ALTER DEFAULT PRIVILEGES FOR ALL ROLES GRANT SELECT ON TABLES TO roacher;
> ```

could we add one more test with `ALTER DEFAULT PRIVILEGES IN SCHEMA xyz GRANT INSERT ON TABLES TO roach;`
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9mVf5wWNeoHO0Q8U:-O8I9mVf5wWNeoHO0Q8V:b-xgfquk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1140)):*
> ```
> statement ok
> CREATE ROLE roachie;
> CREATE ROLE roach;
> ```

nit: roach, roachie, roacher, and roachest are all pretty similar names, so it makes the test a little hard to follow. can we use names like roach_a, roach_b, roach_c, etc.?
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I7VFS8nH8NuhEywkI:-O8I7VFS8nH8NuhEywkJ:b4zcf9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1153)):*
> ```
> 
> statement ok
> SET ROLE admin;
> ```

super nit: the default role in logic tests is `root`, so let's use that here.
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IAGFy9mM95YyS66aM:-O8IAGFy9mM95YyS66aN:b-dakgaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1174)):*
> ```
> WHERE id = (SELECT oid FROM pg_database WHERE datname = 'test2');
> ----
> [
> ```

nit: this is a very large output, which makes it harder to maintain this test (if there's a test diff in a later PR, people are less likely to be able to tell if the diff is valid or not). can we change the assertion to narrow down to the fields we care about? we can use a CTE and `json_array_elements`, like:
```
WITH t AS (
  SELECT json_array_elements(crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor)
    -> 'database'
    -> 'defaultPrivileges'
    -> 'defaultPrivilegesPerRole') AS default_privs_per_role
  FROM system.descriptor
  WHERE id = (SELECT oid FROM pg_database WHERE datname = 'defaultdb')
) SELECT
  default_privs_per_role->'defaultPrivilegesPerObject'->'1'->'users',
  default_privs_per_role->'explicitRole'->'userProto'
FROM t;
```
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1291 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I7r670YMOGHurII-4:-O8I7r670YMOGHurII-5:b-i0clzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1291)):*
> ```
> 
> query ITTTT colnames,rowsort
> SELECT id, database_name, schema_name, obj_name, error FROM crdb_internal.invalid_objects;
> ```

is this the line that's failing under certain logictest configs?

i believe the issue is that there's white space in the results of the query, which confuses the `rowsort` option.

since there's only one result here, we don't need `rowsort`. we can add `WHERE database_name = test2 AND obj_name = ''`  so that there is only ever one result.

another alternative to `rowsort` when we do have multiple results is to add an `ORDER BY` to the query instead of using `rowsort`.

---

also, let's remove `id` from the query. IDs are not deterministic, so including it in the assertion makes the test flaky
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1299 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9Vr3B-6hXU3tPRc7:-O8I9Vr3B-6hXU3tPRc8:bgeei9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1299)):*
> ```
> SELECT * FROM crdb_internal.kv_repairable_catalog_corruptions;
> ----
> parent_id  parent_schema_id  name   id   corruption
> ```

let's remove `id` from the query. IDs are not deterministic, so including it in the assertion makes the test flaky
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9Y0O4ZdKPZH1aI5b:-O8I9Y0O4ZdKPZH1aI5c:bgeei9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1306)):*
> ```
> 
> query ITTTT colnames,rowsort
> SELECT id, database_name, schema_name, obj_name, error FROM crdb_internal.invalid_objects;
> ```

let's remove `id` from the query. IDs are not deterministic, so including it in the assertion makes the test flaky
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8ID013CHN7wWaJsdJ4:-O8ID013CHN7wWaJsdJ5:b-oj1ret) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1317)):*
> ```
>   -> 'defaultPrivilegesPerRole')
> FROM system.descriptor
> WHERE id = (SELECT oid FROM pg_database WHERE datname = 'test2');
> ```

see comment above about using a CTE and json_array_elements

**Review by annrpom — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131782)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 6225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IDIGVAdDio3U7Od7s:-O8IFoiTExHQJyUQk_kl:b-15kp7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/crdb_internal.go#L6225)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

schema descriptors can also have default privileges in them; let's check them here too
</blockquote></details>

done -- ty for pointing that out
___
*[`pkg/sql/catalog/dbdesc/database_desc_builder.go` line 187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IEAZpCQvIsEPiQsCD:-O8IGw3C4wIMhzBf0UE8:b-37mmeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/catalog/dbdesc/database_desc_builder.go#L187)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we'll also need similar logic for schema descriptors
</blockquote></details>

done ('-')7
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9ykp7mDBTekCG6bF:-O8IHizPArhacsaAlCqG:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1121)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: the default role in logic tests is `root`, so let's use that here.
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IEI-I-kdAn0cs60Po:-O8IHuWt06E3KSq6gk-X:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1124)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could we add one more test with `ALTER DEFAULT PRIVILEGES IN SCHEMA xyz GRANT INSERT ON TABLES TO roach;`
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9mVf5wWNeoHO0Q8U:-O8II7eUA3MeWXwEcvxT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1140)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: roach, roachie, roacher, and roachest are all pretty similar names, so it makes the test a little hard to follow. can we use names like roach_a, roach_b, roach_c, etc.?
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I7VFS8nH8NuhEywkI:-O8IIieD0osumLAvjAWr:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1153)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: the default role in logic tests is `root`, so let's use that here.
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8IAGFy9mM95YyS66aM:-O8IIjgF46ORj_NXMXPf:b24a1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1174)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this is a very large output, which makes it harder to maintain this test (if there's a test diff in a later PR, people are less likely to be able to tell if the diff is valid or not). can we change the assertion to narrow down to the fields we care about? we can use a CTE and `json_array_elements`, like:
```
WITH t AS (
  SELECT json_array_elements(crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor)
    -> 'database'
    -> 'defaultPrivileges'
    -> 'defaultPrivilegesPerRole') AS default_privs_per_role
  FROM system.descriptor
  WHERE id = (SELECT oid FROM pg_database WHERE datname = 'defaultdb')
) SELECT
  default_privs_per_role->'defaultPrivilegesPerObject'->'1'->'users',
  default_privs_per_role->'explicitRole'->'userProto'
FROM t;
```
</blockquote></details>

good point! done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1291 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I7r670YMOGHurII-4:-O8IKbTl420pDVTGUhY-:bna6fb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1291)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this the line that's failing under certain logictest configs?

i believe the issue is that there's white space in the results of the query, which confuses the `rowsort` option.

since there's only one result here, we don't need `rowsort`. we can add `WHERE database_name = test2 AND obj_name = ''`  so that there is only ever one result.

another alternative to `rowsort` when we do have multiple results is to add an `ORDER BY` to the query instead of using `rowsort`.

---

also, let's remove `id` from the query. IDs are not deterministic, so including it in the assertion makes the test flaky
</blockquote></details>

done; hm i see i wonder why this only shows up in certain logictest configs
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1299 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9Vr3B-6hXU3tPRc7:-O8IK7IMA0ZqsfdBIiO_:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1299)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

let's remove `id` from the query. IDs are not deterministic, so including it in the assertion makes the test flaky
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8I9Y0O4ZdKPZH1aI5b:-O8IMi4h49XMQfUthMqL:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1306)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

let's remove `id` from the query. IDs are not deterministic, so including it in the assertion makes the test flaky
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges` line 1317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-O8ID013CHN7wWaJsdJ4:-O8IMilW0vkj9i92jzWq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1698ee2280c8da8cdcaccb0dc4b30f76671a1adf/pkg/sql/logictest/testdata/logic_test/crdb_internal_default_privileges#L1317)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

see comment above about using a CTE and json_array_elements
</blockquote></details>

done

**Comment by annrpom** [comment]
ah right let me remove the rowsort option

**Comment by annrpom** [comment]
oki RFAL @rafiss

**Review by rafiss — APPROVED** [review]
nice work!

> Reviewed 1 of 2 files at r1, 1 of 3 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131782#-:-O8cdhUa2gMC5giCIyTJ:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

**Comment by annrpom** [comment]
TFTR! ('-')7 

bors r=rafiss

**Comment by rafiss** [comment]
I believe we were only seeing this issue in Cloud clusters on 24.1 and later, so if the backport to 23.2 and earlier has annoying merge conflicts, we can probably skip those.

**Comment by annrpom** [comment]
i think you sent that at the same time that i put that backport up @rafiss; no annoying merge conflict; it's just this line of code that we didnt have in 23.2 and below:

https://github.com/cockroachdb/cockroach/blob/3717e00ae1ff447014e32c483a0e38f396097ba3/pkg/sql/catalog/dbdesc/database_desc_builder.go#L195-L199

**Comment by annrpom** [comment]
ah but 23.1 is a diff story; i'll leave that one alone @rafiss
